lib/Reader/readerir.cpp

```diff
@@ -5994,8 +5994,35 @@ void GenIR::switchOpcode(IRNode *Opr) {
   BasicBlock *CurrBlock = LLVMBuilder->GetInsertBlock();
   TerminatorInst *TermInst = CurrBlock->getTerminator();
   SwitchInst *SwitchInstruction = cast<SwitchInst>(TermInst);
-
   SwitchInstruction->setCondition(Opr);
+
+  Type *OprType = Opr->getType();
+  LLVMContext &LLVMContext = *this->JitContext->LLVMContext;
+  Type *Int32Ty = Type::getInt32Ty(LLVMContext);
+  if (OprType != Int32Ty) {
+    // When we made the case values we assumed int32 as the type for the
+    // values. But now we have an operand of a different type. LLVM
+    // requires the values and the switch value to have the same type.
+    // So change the type of the values to be the type of the selector.
+    IntegerType *OprIntType = cast<IntegerType>(OprType);
+
+    // If a small selector type was used, e.g. i8, the type might not
+    // be big enough to hold all the case values. But the value for the
+    // selector is popped off the evaluation stack, so it should be at
+    // least 32 bits. So check that.
+    unsigned OprBitWidth = OprIntType->getBitWidth();
+    assert((OprBitWidth > 32) && "Selector bit width is less then 32");
+
+    // Create new values of the correct type.
+    int NumCases = SwitchInstruction->getNumCases();
+    SwitchInst::CaseIt Case = SwitchInstruction->case_begin();
+    for (size_t I = 0; I < NumCases; I++) {
+      ConstantInt *Value = ConstantInt::get(
+          *JitContext->LLVMContext, APInt(OprBitWidth, (uint64_t)I, false));
```

Comment:
Currently we unconditionally make all the case values i32. The current code could fail if the selector was i32 and original case values had some type other than i32. We could instead check the type of the first case and compare that with the type of the selector. If they differed then we could do the fix. Adding an assert, as you suggest, would at least detect the problem.